File: sdks/node/packages/plugin-sdk/src/manifest-builder.ts
// SPDX-FileCopyrightText: 2024-2026 Temps Contributors
// SPDX-License-Identifier: MIT OR Apache-2.0

/**
 * Builder pattern for constructing PluginManifest instances.
 *
 * Mirrors the Rust `PluginManifest::builder()` pattern.
 */

import type { NavEntry, NavSection, PluginManifest, UiManifest } from "./types.js";

export class ManifestBuilder {
  private _name: string;
  private _version: string;
  private _displayName?: string;
  private _description?: string;
  private _nav: NavEntry[] = [];
  private _ui?: UiManifest;
  private _requiresDb = false;
  private _healthPath = "/health";
  private _events: string[] = [];

  constructor(name: string, version: string) {
    this._name = name;
    this._version = version;
  }

  displayName(name: string): this {
    this._displayName = name;
    return this;
  }

  description(desc: string): this {
    this._description = desc;
    return this;
  }

  nav(entry: NavEntry): this {
    this._nav.push(entry);
    return this;
  }

  /**
   * Add a navigation entry using shorthand parameters.
   */
  addNav(
    label: string,
    icon: string,
    path: string,
    options?: { section?: NavSection; order?: number }
  ): this {
    this._nav.push({
      label,
      icon,
      path,
      section: options?.section ?? "platform",
      order: options?.order ?? 50,
    });
    return this;
  }

  ui(manifest: UiManifest): this {
    this._ui = manifest;
    return this;
  }

  requiresDb(requires: boolean): this {
    this._requiresDb = requires;
    return this;
  }

  healthPath(path: string): this {
    this._healthPath = path;
    return this;
  }

  event(eventType: string): this {
    this._events.push(eventType);
    return this;
  }

  events(eventTypes: string[]): this {
    this._events.push(...eventTypes);
    return this;
  }

  build(): PluginManifest {
    return {
      name: this._name,
      version: this._version,
      display_name: this._displayName,
      description: this._description,
      nav: this._nav,
      ui: this._ui,
      requires_db: this._requiresDb,
      health_path: this._healthPath,
      events: this._events,
    };
  }
}

/**
 * Create a new manifest builder.
 *
 * @example
 * ```ts
 * const manifest = createManifest("my-plugin", "0.1.0")
 *   .displayName("My Plugin")
 *   .description("Does amazing things")
 *   .addNav("Dashboard", "layout-dashboard", "/dashboard")
 *   .event("deployment.succeeded")
 *   .build();
 * ```
 */
export function createManifest(
  name: string,
  version: string
): ManifestBuilder {
  return new ManifestBuilder(name, version);
}

File: sdks/python/temps_plugin/manifest_builder.py
# SPDX-FileCopyrightText: 2024-2026 Temps Contributors
# SPDX-License-Identifier: MIT OR Apache-2.0

"""Fluent manifest builder.

Mirrors: sdks/node/packages/plugin-sdk/src/manifest-builder.ts
"""

from __future__ import annotations

from temps_plugin.types import NavEntry, NavSection, PluginManifest


class ManifestBuilder:
    """Fluent builder for PluginManifest."""

    def __init__(self, name: str, version: str) -> None:
        self._manifest = PluginManifest(name=name, version=version)

    def display_name(self, name: str) -> ManifestBuilder:
        self._manifest.display_name = name
        return self

    def description(self, desc: str) -> ManifestBuilder:
        self._manifest.description = desc
        return self

    def requires_db(self, val: bool = True) -> ManifestBuilder:
        self._manifest.requires_db = val
        return self

    def health_path(self, path: str) -> ManifestBuilder:
        self._manifest.health_path = path
        return self

    def add_nav(
        self,
        label: str,
        icon: str,
        path: str,
        section: NavSection = NavSection.PLATFORM,
        order: int = 50,
    ) -> ManifestBuilder:
        self._manifest.nav.append(
            NavEntry(label=label, icon=icon, path=path, section=section, order=order)
        )
        return self

    def event(self, event_type: str) -> ManifestBuilder:
        self._manifest.events.append(event_type)
        return self

    def build(self) -> PluginManifest:
        return self._manifest


def create_manifest(name: str, version: str) -> ManifestBuilder:
    """Create a new ManifestBuilder."""
    return ManifestBuilder(name, version)
